storage/s3_test.go

```diff
@@ -266,44 +267,134 @@ func TestS3ListContextCancelled(t *testing.T) {
+		},
 
-	sess := unit.Session
-	sess.Config.Retryer = newCustomRetryer(expectedRetry)
+		// Throttling errors
+		{
+			name: "ProvisionedThroughputExceededException",
+			err:  awserr.New("ProvisionedThroughputExceededException", "provisioned throughput exceeded exception", nil),
+		},
+		{
+			name: "Throttling",
+			err:  awserr.New("Throttling", "throttling", nil),
+		},
+		{
+			name: "ThrottlingException",
+			err:  awserr.New("ThrottlingException", "throttling exception", nil),
+		},
+		{
+			name: "RequestLimitExceeded",
+			err:  awserr.New("RequestLimitExceeded", "request limit exceeded", nil),
+		},
+		{
+			name: "RequestThrottled",
+			err:  awserr.New("RequestThrottled", "request throttled", nil),
+		},
+		{
+			name: "RequestThrottledException",
+			err:  awserr.New("RequestThrottledException", "request throttled exception", nil),
+		},
 
-	mockApi := s3.New(sess)
-	mockS3 := &S3{
-		api: mockApi,
+		// Expired credential errors
+		{
+			name: "ExpiredToken",
+			err:  awserr.New("ExpiredToken", "expired token", nil),
+		},
+		{
+			name: "ExpiredTokenException",
+			err:  awserr.New("ExpiredTokenException", "expired token exception", nil),
+		},
+
+		// Connection errors
+		{
+			name: "connection reset",
+			err:  fmt.Errorf("connection reset by peer"),
+		},
+		{
+			name: "broken pipe",
+			err:  fmt.Errorf("broken pipe"),
+		},
+
+		// Unknown errors
+		{
+			name: "an unknown error is also retried by SDK",
+			err:  fmt.Errorf("an error that is not known to the SDK"),
+		},
 	}
 
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
+	url, err := url.New("s3://bucket/key")
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
 
-	mockApi.Handlers.Send.Clear() // mock sending
-	mockApi.Handlers.Unmarshal.Clear()
-	mockApi.Handlers.UnmarshalMeta.Clear()
-	mockApi.Handlers.ValidateResponse.Clear()
-	mockApi.Handlers.Unmarshal.PushBack(func(r *request.Request) {
-		r.Error = awserr.New("InternalError", "s3 api failed", nil)
-	})
+	const expectedRetry = 5
+	for _, tc := range testcases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			sess := unit.Session
+			sess.Config.Retryer = newCustomRetryer(expectedRetry)
 
-	retried := -1
-	// Add a request handler to the AfterRetry handler stack that is used by the
-	// SDK to be executed after the SDK has determined if it will retry.
-	mockApi.Handlers.AfterRetry.PushBack(func(_ *request.Request) {
-		retried++
-	})
+			mockApi := s3.New(sess)
+			mockS3 := &S3{
+				api: mockApi,
+			}
 
-	for range mockS3.List(ctx, url, true) {
-	}
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			mockApi.Handlers.Send.Clear() // mock sending
+			mockApi.Handlers.Unmarshal.Clear()
+			mockApi.Handlers.UnmarshalMeta.Clear()
+			mockApi.Handlers.ValidateResponse.Clear()
+			mockApi.Handlers.Unmarshal.PushBack(func(r *request.Request) {
+				r.Error = tc.err
+				r.HTTPResponse = &http.Response{StatusCode: 500}
```

Comment:
Not all errors listed above return 500 according to doc (https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html).

Reply:
It's set to make aws defaultretryer happy. Otherwise, there are many status codes that we need to track. I'm going to remove that code now.